File: strings/FirstUniqueChar.java
package strings;

import java.util.*;

public class FirstUniqueChar {
    
    public static void main(String[] args) {
	String s1 = "abacb";
	String s2 = "";
	String s3 = null;
	String s4 = "a";
	String s5 = "aabbccc";
	String s6 = "abbccc";
	System.out.println("HashMap----");
	System.out.println(findHashMap(s1) == 'c');
	System.out.println(findHashMap(s2) == null);
	System.out.println(findHashMap(s3) == null);
	System.out.println(findHashMap(s4) == 'a');
	System.out.println(findHashMap(s5) == null);
	System.out.println(findHashMap(s6) == 'a');
	System.out.println("Double Loop----");
	System.out.println(findDoubleLoop(s1) == 'c');
	System.out.println(findDoubleLoop(s2) == null);
	System.out.println(findDoubleLoop(s3) == null);
	System.out.println(findDoubleLoop(s4) == 'a');
	System.out.println(findDoubleLoop(s5) == null);
	System.out.println(findDoubleLoop(s6) == 'a');
	System.out.println("Sorted----");
	System.out.println(findSorted(s1) == 'c');
	System.out.println(findSorted(s2) == null);
	System.out.println(findSorted(s3) == null);
	System.out.println(findSorted(s4) == 'a');
	System.out.println(findSorted(s5) == null);
	System.out.println(findSorted(s6) == 'a');

    }
    

    // Log Linear Time - O(n Log n)
    // aadee
    public static Character findSorted(String str) {
	if (str == null || str.length() == 0) {
	    return null;
	} else if (str.length() == 1) {
	    return str.charAt(0);
	}
	Character[] charArr = new Character[str.length()];
	for (int i=0; i<str.length(); i++) {
	    charArr[i] = str.charAt(i);
	}
	Arrays.sort(charArr);
	int i=0;
	while (i<charArr.length) {
	    char cur = charArr[i];
	    if (i == charArr.length-1 || charArr[i+1] != cur) {
		return charArr[i];
	    }
	    while (i<charArr.length && charArr[i] == cur) {
		i++;
	    }
	}
	return null;
	
    }

    // Quadratic Time - O(n^2)
    public static Character findDoubleLoop(String str) {
	if (str == null || str.length() == 0) {
	    return null;
	}
	for (int i=0; i<str.length(); i++) {
	    int j = 0;
	    boolean foundDupe = false;
	    while (!foundDupe && j<str.length()) {
		if (str.charAt(i) == str.charAt(j) && i != j) {   //need to check if already tried this char
		    foundDupe = true;
		} else {
		    j++;
		}
	    }
	    if (!foundDupe) {
		return str.charAt(i);
	    }
	}
	return null;	   
    }

    // Linear Time - O(n), Space O(n)
    public static Character findHashMap(String str) {
	if (str == null || str.length() == 0) {
	    return null;
	} else if (str.length() == 1) {
	    return str.charAt(0);
	}
	// Build HashMap to store count of each char
	Map<Character,Integer> map = new HashMap();
	for (int i=0; i<str.length(); i++) {
	    if ( map.containsKey(str.charAt(i)) ) {
		map.put(str.charAt(i),map.get(str.charAt(i))+1);
	    } else {
		map.put(str.charAt(i),1);
	    }
	}

	// Loop through initial str, checking if count == 1
	for (int i=0; i<str.length(); i++) {
	    if (map.get(str.charAt(i)) == 1) {
		return str.charAt(i);
	    }
	}

	return null; // no unique characters

    }


}

File: strings/python/first_unique_char.py
"""
[FindFirstUniqueChar]
Return the first non-repeated character in a string. 
If not found, return null. Assume the string is NOT sorted. 
"""

#Cases
"""
1) Empty or None 
2) No repeats
3) Single repeat 
4) Multiple repeats (return first found?)
5) 1,2,3 length strings
"""

#Approaches
"""
1) Sort string O(n log n), then loop O(n) and check if str1[i] == str1[i+1] (Won't work for first unique)
2) HashMap to store unique chars. Loop O(n) check if in HashMap, if not, store.
3) Naive Double Loop O(n^2), For each, for each, see if another exists. Return first found.
4) Can we do linear time?
"""

def get_first_unique_char_loops(str1):
	i = 0
	while i < len(str1):
		j = i+1
		while j < len(str1):
			if str1[i] == str1[j]:
				return str1[i]
			j+=1
		i+=1
	return None

def get_first_unique_char_dict(str1):
	chars = {}
	for s in str1:
		if chars.get(s) != None:
			return s
		chars[s] = s
	return None


#Tests

def test_get_first_unique_char_loops():
	assert get_first_unique_char_loops("ABCBD") == "B"
	assert get_first_unique_char_loops("") == None
	assert get_first_unique_char_loops("A") == None
	assert get_first_unique_char_loops("AB") == None
	assert get_first_unique_char_loops("ABCC") == "C"

def test_get_first_unique_char_dict():
	assert get_first_unique_char_dict("ABCBD") == "B"
	assert get_first_unique_char_dict("") == None
	assert get_first_unique_char_dict("A") == None
	assert get_first_unique_char_dict("AB") == None
	assert get_first_unique_char_dict("ABCC") == "C"

if __name__ == "__main__":
	test_get_first_unique_char_loops()
	test_get_first_unique_char_dict()


